addons/package_hierarchy/models/stock_quant_package.py

```diff
@@ -39,12 +39,47 @@ def _check_package_recursion(self):
 
     def _check_not_multi_location(self):
         if not self.env.context.get("bypass_multi_location_check", False):
-            for package in self:
-                locations = package.children_quant_ids.mapped('location_id')
-                if len(locations) > 1:
-                    raise ValidationError(_('Package cannot be in multiple '
-                                            'locations:\n%s\n%s') % (package.name,
-                                                                    ', '.join( [l.name for l in locations])))
+            packages = self.filtered(lambda p: p.package_id)
+            parent_packages = self.filtered(lambda p: not p.package_id)
```

Comment:
As we're making these changes around performance, we could just do:

`parent_packages = self - packages`

Although not sure the gain would be noticeable, also makes the code a bit less explicit.

Just thinking out loud. :)